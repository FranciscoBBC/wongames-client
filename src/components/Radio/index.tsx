import React, { InputHTMLAttributes } from 'react'
import * as S from './styles'

type RadioValue = string | ReadonlyArray<string> | number

export type RadioProps = {
  label?: string
  labelFor?: string
  labelColor?: 'white' | 'black'
  onCheck?: (value: RadioValue) => void
  value?: RadioValue
} & InputHTMLAttributes<HTMLInputElement>

const Radio = ({
  label,
  labelFor = '',
  labelColor = 'white',
  onCheck,
  value,
  ...props
}: RadioProps) => {
  const onChange = () => {
    !!onCheck && onCheck(value!)
  }

  return (
    <S.Wrapper>
      <S.Input
        type="radio"
        id={labelFor}
        onChange={onChange}
        value={value}
        {...props}
      />
      {!!labelFor && (
        <S.Label htmlFor={labelFor} labelColor={labelColor}>
          {label}
        </S.Label>
      )}
    </S.Wrapper>
  )
}

export default Radio
