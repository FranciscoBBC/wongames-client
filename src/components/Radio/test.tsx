import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'
import Radio from '.'

describe('<Radio />', () => {
  it('should render a white label', () => {
    const { container } = renderWithTheme(
      <Radio label="label radio" labelFor="check" value="radio value" />
    )

    expect(screen.getByLabelText(/label radio/i)).toBeInTheDocument()
    expect(screen.getByText(/label radio/i)).toBeInTheDocument()
    expect(screen.getByText(/label radio/i)).toHaveStyle({ color: '#FAFAFA' })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a black label', () => {
    renderWithTheme(
      <Radio label="label radio" labelFor="check" labelColor="black" />
    )

    expect(screen.getByText(/label radio/i)).toBeInTheDocument()
    expect(screen.getByText(/label radio/i)).toHaveStyle({ color: '#030517' })
  })

  it('should render without a label', () => {
    renderWithTheme(<Radio />)

    expect(screen.getByRole('radio')).toBeInTheDocument()
    expect(screen.queryByLabelText(/label radio/i)).not.toBeInTheDocument()
  })

  it('should dispatch onCheck when label status changes', async () => {
    const onCheck = jest.fn()

    renderWithTheme(
      <Radio
        label="label radio"
        labelFor="check"
        value="radio value"
        onCheck={onCheck}
      />
    )

    expect(onCheck).not.toHaveBeenCalled()

    userEvent.click(screen.getByLabelText(/label radio/i))

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1)
    })

    expect(onCheck).toHaveBeenCalledWith('radio value')
  })

  it('Should be accessible with tab', () => {
    renderWithTheme(
      <Radio label="label radio" labelFor="check" value="radio value" />
    )

    expect(document.body).toHaveFocus()
    userEvent.tab()
    expect(screen.getByLabelText(/label radio/i)).toHaveFocus()
  })
})
