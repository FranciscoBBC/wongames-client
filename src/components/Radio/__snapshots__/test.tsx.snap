// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render a white label 1`] = ` `;
